### PR TITLE
depositors query

### PR DESCRIPTION
**Pull request by rvanheest**
Fixes EASY-

#### When applied it will
* add a `depositors` query

#### Still to do
* [x] change documentation
* [x] ~time based filters~ (_too difficult for now, maybe later once we take an annotation based approach_)
* [x] tests for `QueryGenerator.searchDepositors`

#### By submitting this pull request I confirm that
* [x] all files contain licenses (`mvn license:format`)
* [x] the project compiles (`mvn clean install`)
* [x] all unit tests pass (`mvn test`)
* [x] the changes have been tested on `deasy`

@DANS-KNAW/easy for review

### CODE REVIEW

File: src/main/scala/nl.knaw.dans.easy.properties/app/repository/sql/QueryGenerator.scala [hunk 1]
```diff
@@ -30,38 +30,39 @@ object QueryGenerator {
...
-      (tableName, query, setString(labelValue(filter)) :: Nil)
+  private type TableName = String
+  private type KeyName = String
+  private type LabelName = String
+  private type Query = String
+
+  private def createSearchSubQuery[T <: DepositFilter](filter: T)(tableName: TableName, labelName: LabelName, labelValue: T => String): (TableName, Query, List[PrepStatementResolver]) = {
+    val query = filter.filter match {
+      case SeriesFilter.ALL =>
+        s"SELECT DISTINCT depositId FROM $tableName WHERE $labelName = ?"
```
Comment (jo-pol) [inline src/main/scala/nl.knaw.dans.easy.properties/app/repository/sql/QueryGenerator.scala]: Not new but perhaps: labelName -> columnName?

File: src/main/scala/nl.knaw.dans.easy.properties/app/repository/sql/QueryGenerator.scala [hunk 2]
```diff
@@ -107,6 +108,54 @@ object QueryGenerator {
...
+    )
+      .collect {
+        case Some((tableName, q, values)) if queryWherePart.isEmpty => s"INNER JOIN ($q) AS ${ tableName }SearchResult ON Deposit.depositId = ${ tableName }SearchResult.depositId" -> values
+        case Some((tableName, q, values)) => s"INNER JOIN ($q) AS ${ tableName }SearchResult ON SelectedDeposits.depositId = ${ tableName }SearchResult.depositId" -> values
+      }
+      .foldLeft(("", List.empty[PrepStatementResolver])) {
+        case (("", vs), (subQuery, values)) => subQuery -> (values ::: vs)
+        case ((q, vs), (subQuery, values)) => s"$q $subQuery" -> (values ::: vs)
+      }
+
```
Comment (jo-pol) [inline src/main/scala/nl.knaw.dans.easy.properties/app/repository/sql/QueryGenerator.scala:142]: I thought I started to understand...

File: src/test/scala/nl.knaw.dans.easy.properties/app/repository/sql/QueryGeneratorSpec.scala [hunk 3]
```diff
@@ -377,6 +380,248 @@ class QueryGeneratorSpec extends TestSupportFixture with MockFactory {
...
+        |  SELECT State.depositId
+        |  FROM State
+        |  INNER JOIN (
+        |    SELECT depositId, max(timestamp) AS max_timestamp
+        |    FROM State
+        |    GROUP BY depositId
+        |  ) AS StateWithMaxTimestamp
+        |  ON State.timestamp = StateWithMaxTimestamp.max_timestamp
+        |  WHERE label = ?
+        |) AS StateSearchResult
```
Comment (jo-pol) [inline src/test/scala/nl.knaw.dans.easy.properties/app/repository/sql/QueryGeneratorSpec.scala:565]: I wonder we could use a derived table (maintained with a stored procedure?) or view with the last state of a deposit. The view would reduce complexity of queries, the derived table would increase maintenance performance but reduce query performance.

Comment (jo-pol) [reply src/test/scala/nl.knaw.dans.easy.properties/app/repository/sql/QueryGeneratorSpec.scala:565]:
```sql
CREATE VIEW LastState AS
SELECT State.stateId, State.depositId, State.label, State.description, State.timestamp 
FROM State
INNER JOIN (
  SELECT depositId, max(timestamp) AS max_timestamp
  FROM State
  GROUP BY depositId
) AS StateWithMaxTimestamp
ON State.timestamp = StateWithMaxTimestamp.max_timestamp
```

same for SimpleProperties

Comment (rvanheest) [reply src/test/scala/nl.knaw.dans.easy.properties/app/repository/sql/QueryGeneratorSpec.scala:565]: Great idea! As discussed offline, we'll do this in another issue (https://drivenbydata.atlassian.net/browse/EASY-2415)

**Pull request merged**